File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.java
class Solution {
    public int minCost(String s, int[] arr) {
        int ans=0;
        int i=0;
        while(i<s.length()){
            if(i+1<s.length() && s.charAt(i)==s.charAt(i+1)){
                int j=i;
                int maxValue=0;
                while(j<s.length() && s.charAt(j)==s.charAt(i)){
                    if(arr[j]>maxValue){
                        maxValue=arr[j];
                    }
                    j++;
                }
                while(i<j){
                    ans+=arr[i];
                    i++;
                }
                ans-=maxValue;
            }else{
                i++;
            }
        }
        return ans;
    }
}

File: Leetcode/1578 Minimum Time to Make Rope Colorful/1578 Minimum Time to Make Rope Colorful.cpp
//There are two approch for this Problem both approch are optimal also both uses O(n) Time Complexity

#include<bits/stdc++.h>
using namespace std;


//===========================Approch 1======================


class Solution {
public:
    int minCost(string s, vector<int>& time) {
        int ans = 0;
        int n = s.length();
        for(int i = 1; i < n; i++)
        {
            if(s[i] == s[i - 1])
            {
                ans += min(time[i], time[i - 1]); 
                time[i] = max(time[i], time[i - 1]); 
            }     
        }
        
        return ans;
    }
};


//===========================Approch 2 =========================




class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
//         initializing 2 pointers
        int totalTime =0, i =0, j =0;
        while(i<neededTime.size() && j < neededTime.size()){
            int currTotal =0, currMax =0;
//             find all balloons having the same color as the balloon indexed[i], record the total removal time & maxi removal time
            while(j < neededTime.size() && colors[i]== colors[j]){
                currTotal+= neededTime[j];
                currMax = max(currMax, neededTime[j]);
                j++;
            }
//             once we reach the end of the curr group, add the cost of this group to total_time, & reset 2 pointers
            totalTime+=currTotal-currMax;
            i=j;
        }
        return totalTime;
//         TC: O(n), SC: O(1)
    }
};
